File: api/app/barriers.py
"""Deterministic Phase 7 barrier placement and road-edge blocking.

A barrier is not decoration: it is bound to one stable road edge id and blocks that
edge in both directions, so the router, the distance matrices and the published plan
all treat the cut the same way. ``blockedEdgeIds`` stays the single source of truth and
is always derived from the active barriers, so a barrier and the set of blocked edges
can never disagree.

Units stay explicit: points are metres in the local Three.js XZ plane, distances are
metres and headings are compass-style degrees (0 points to +z, 90 points to +x), exactly
like the claw snap of Phase 6 that this module mirrors.
"""

from __future__ import annotations

from dataclasses import dataclass
from math import hypot
from typing import Any, Final, Iterable

from .simulation import EPSILON_M, heading_degrees

# Barrier snap radius; mirrors `SNAP_EDGE_MAX_RADIUS_M` in `frontend/src/city/dataset.ts`.
SNAP_EDGE_MAX_RADIUS_M: Final = 12.0
# Frozen MVP limit: three simultaneous active barriers.
MAX_BARRIERS: Final = 3


@dataclass(frozen=True, slots=True)
class EdgeSnap:
    """Closest road edge inside the barrier radius."""

    edge_id: str
    from_node_id: str
    to_node_id: str
    bidirectional: bool
    projected_point: dict[str, float]
    t: float
    distance_meters: float
    heading_degrees: float


def project_on_segment(
    point: dict[str, float], start: dict[str, float], end: dict[str, float]
) -> tuple[dict[str, float], float, float]:
    """Project ``point`` on the XZ segment ``start``-``end``, with ``t`` clamped.

    A degenerate segment (both endpoints on the same XZ point) resolves to ``t = 0``
    instead of dividing by zero.
    """
    dx = end["x"] - start["x"]
    dz = end["z"] - start["z"]
    length_squared = dx * dx + dz * dz
    if length_squared <= EPSILON_M:
        return (
            {"x": start["x"], "y": start["y"], "z": start["z"]},
            0.0,
            hypot(point["x"] - start["x"], point["z"] - start["z"]),
        )
    raw_t = ((point["x"] - start["x"]) * dx + (point["z"] - start["z"]) * dz) / length_squared
    t = min(1.0, max(0.0, raw_t))
    projected_point = {
        "x": start["x"] + t * dx,
        "y": start["y"] + t * (end["y"] - start["y"]),
        "z": start["z"] + t * dz,
    }
    distance = hypot(point["x"] - projected_point["x"], point["z"] - projected_point["z"])
    return projected_point, t, distance


def _edges(graph: dict[str, Any]) -> list[dict[str, Any]]:
    return sorted(graph["edges"], key=lambda edge: edge["edgeId"])


def _nodes(graph: dict[str, Any]) -> dict[str, dict[str, Any]]:
    return {node["nodeId"]: node for node in graph["nodes"]}


def edge_snap(graph: dict[str, Any], edge_id: str) -> EdgeSnap | None:
    """Snap of an explicit edge id, at its midpoint, or ``None`` when it is unknown.

    The caller asked for the whole edge, not for a point on it, so the reported point is
    the edge midpoint and the distance is zero: the barrier really is on the road.
    """
    edge = next((item for item in graph["edges"] if item["edgeId"] == edge_id), None)
    if edge is None:
        return None
    nodes = _nodes(graph)
    start = nodes.get(edge["fromNodeId"])
    end = nodes.get(edge["toNodeId"])
    if start is None or end is None:
        return None
    start_position, end_position = start["position"], end["position"]
    midpoint = {
        "x": (start_position["x"] + end_position["x"]) / 2,
        "y": (start_position["y"] + end_position["y"]) / 2,
        "z": (start_position["z"] + end_position["z"]) / 2,
    }
    return EdgeSnap(
        edge_id=edge["edgeId"],
        from_node_id=edge["fromNodeId"],
        to_node_id=edge["toNodeId"],
        bidirectional=bool(edge.get("bidirectional", False)),
        projected_point=midpoint,
        t=0.5,
        distance_meters=0.0,
        heading_degrees=heading_degrees(start_position, end_position),
    )


def nearest_edge(
    graph: dict[str, Any],
    point: dict[str, float],
    max_radius: float = SNAP_EDGE_MAX_RADIUS_M,
    excluded_edge_ids: Iterable[str] = (),
) -> EdgeSnap | None:
    """Closest road edge inside ``max_radius``, or ``None`` outside it.

    Already blocked edges are excluded before the distance test, which is what stops a
    barrier from being placed twice on the same road. An exact tie (within ``EPSILON_M``)
    always goes to the lexicographically smaller edge id, so the same drop resolves to
    the same road on every run.
    """
    excluded = set(excluded_edge_ids)
    nodes = _nodes(graph)
    best: EdgeSnap | None = None
    for edge in _edges(graph):
        if edge["edgeId"] in excluded:
            continue
        start = nodes.get(edge["fromNodeId"])
        end = nodes.get(edge["toNodeId"])
        if start is None or end is None:
            continue
        projected_point, t, distance = project_on_segment(
            point, start["position"], end["position"]
        )
        if distance > max_radius + EPSILON_M:
            continue
        if best is None or distance < best.distance_meters - EPSILON_M:
            best = EdgeSnap(
                edge_id=edge["edgeId"],
                from_node_id=edge["fromNodeId"],
                to_node_id=edge["toNodeId"],
                bidirectional=bool(edge.get("bidirectional", False)),
                projected_point=projected_point,
                t=t,
                distance_meters=distance,
                heading_degrees=heading_degrees(start["position"], end["position"]),
            )
            continue
        is_tie = abs(distance - best.distance_meters) <= EPSILON_M
        if is_tie and edge["edgeId"] < best.edge_id:
            best = EdgeSnap(
                edge_id=edge["edgeId"],
                from_node_id=edge["fromNodeId"],
                to_node_id=edge["toNodeId"],
                bidirectional=bool(edge.get("bidirectional", False)),
                projected_point=projected_point,
                t=t,
                distance_meters=distance,
                heading_degrees=heading_degrees(start["position"], end["position"]),
            )
    return best


def blocked_edge_ids(barriers: Iterable[dict[str, Any]]) -> list[str]:
    """Derive ``blockedEdgeIds`` from the active barriers, in a stable order."""
    return sorted({barrier["blockedEdgeId"] for barrier in barriers})


def barrier_id_for(sequence: int) -> str:
    """Barrier id for one placement sequence number, e.g. ``B-2``.

    Ids are handed out in increasing order and never reused inside a scenario, so a
    removed barrier can never come back as a different road closure.
    """
    return f"B-{int(sequence)}"

File: frontend/src/scenario/barriers.ts
/**
 * Phase 7 barrier placement, preview and impact, as pure functions.
 *
 * The barrier tool is the drag counterpart of the Phase 6 claw, and it follows the same
 * rule: a pointer move only re-resolves the nearest road edge, and nothing here plans a
 * route. Placement, removal and the before/after comparison belong to the published
 * snapshot, which is why the scenario recomputes exactly once per accepted barrier.
 *
 * A barrier blocks a `blockedEdgeId`, never a pixel position, and it blocks that edge in
 * both directions: every MVP road is bidirectional.
 */

import {
  SNAP_EDGE_MAX_RADIUS_M,
  createRoadNetwork,
  roadWidthMetersFor,
  type CityPoint,
  type RoadNetwork,
} from '../city/dataset';
import { edgeDirection, headingDegrees, nearestRoadEdge } from '../city/selection';

import type { ScenarioSnapshot } from './scenario';
import { currentRoutePlan } from './simulation';

/** Reason reported when no road edge is close enough to receive a barrier. */
export type BarrierRejectionCode = 'SNAP_NO_VALID_EDGE';

export interface BarrierPreview {
  accepted: boolean;
  edgeId: string | null;
  /** Raw world point under the pointer, kept so the marker stays where the user dragged. */
  pointer: CityPoint;
  /** Where the barrier post would be placed: the projection on the candidate edge. */
  projectedPoint: CityPoint;
  /** Centre of the candidate edge, where the road highlight is drawn. */
  edgeMidpoint: CityPoint;
  edgeLengthMeters: number;
  edgeWidthMeters: number;
  distanceMeters: number;
  /** Edge heading in degrees; the barrier arm is rotated to cross the road. */
  headingDegrees: number;
  reason: BarrierRejectionCode | null;
}

export interface BarrierPlacement {
  barrierId: string;
  edgeId: string;
  /** Snapped position from the published barrier, in local world metres. */
  position: CityPoint;
  headingDegrees: number;
  selected: boolean;
}

export type BarrierVehicleReason = 'CLOSED_ROAD' | 'NO_WORK' | 'DELAYED_ORDER';

export interface AffectedVehicle {
  vehicleId: string;
  reason: BarrierVehicleReason;
}

export interface AffectedOrder {
  orderId: string;
  reason: string;
}

/**
 * What the active road closures cost, as far as one snapshot can tell.
 *
 * The rules are deliberately conservative:
 * - an order the plan could not assign is reported with the plan's own reason;
 * - an order the plan delays is reported as delayed;
 * - a vehicle whose published route still crosses a closed road is reported, because
 *   that is exactly the invariant Phase 7 has to make visible;
 * - a vehicle with no stops while the plan dropped orders is reported as idle work;
 * - a vehicle carrying a delayed order is reported with it.
 *
 * A closure that changes nothing highlights nothing, which is the honest answer: there
 * is no affected vehicle or order to point at.
 */
export function barrierImpact(snapshot: ScenarioSnapshot): {
  active: boolean;
  edgeIds: string[];
  vehicles: AffectedVehicle[];
  orders: AffectedOrder[];
} {
  const edgeIds = [...new Set(snapshot.blockedEdgeIds)].sort();
  if (snapshot.barriers.length === 0 || edgeIds.length === 0) {
    return { active: false, edgeIds: [], vehicles: [], orders: [] };
  }

  const blocked = new Set(edgeIds);
  const plan = currentRoutePlan(snapshot);
  const vehicles = new Map<string, BarrierVehicleReason>();
  const orders = new Map<string, string>();
  if (plan) {
    for (const item of plan.unassignedOrders) orders.set(item.orderId, item.reason);
    const hasDroppedOrders = plan.unassignedOrders.length > 0;
    for (const route of plan.vehicles) {
      if (route.edgeSequence.some((edgeId) => blocked.has(edgeId))) {
        vehicles.set(route.vehicleId, 'CLOSED_ROAD');
      } else if (route.stops.length === 0 && hasDroppedOrders) {
        vehicles.set(route.vehicleId, 'NO_WORK');
      }
      for (const stop of route.stops) {
        if (stop.delaySeconds <= 0) continue;
        if (!orders.has(stop.orderId)) orders.set(stop.orderId, 'DELAYED');
        if (!vehicles.has(route.vehicleId)) vehicles.set(route.vehicleId, 'DELAYED_ORDER');
      }
    }
  }

  return {
    active: true,
    edgeIds,
    vehicles: [...vehicles.entries()]
      .map(([vehicleId, reason]) => ({ vehicleId, reason }))
      .sort((left, right) => (left.vehicleId < right.vehicleId ? -1 : 1)),
    orders: [...orders.entries()]
      .map(([orderId, reason]) => ({ orderId, reason }))
      .sort((left, right) => (left.orderId < right.orderId ? -1 : 1)),
  };
}

/** The set of vehicle ids to highlight for the active closures. */
export function affectedVehicleIds(snapshot: ScenarioSnapshot): Set<string> {
  return new Set(barrierImpact(snapshot).vehicles.map((item) => item.vehicleId));
}

/**
 * Resolve one drop preview for a pointer position.
 *
 * The already blocked roads are excluded before the distance test, exactly like the
 * server does, so the preview can never offer a road the server would refuse.
 */
export function resolveBarrierPreview(
  network: RoadNetwork,
  pointer: CityPoint,
  excludedEdgeIds: Iterable<string> = [],
): BarrierPreview {
  const snap = nearestRoadEdge(network, pointer, {
    maxRadius: SNAP_EDGE_MAX_RADIUS_M,
    excludedEdgeIds,
  });
  if (!snap) {
    return {
      accepted: false,
      edgeId: null,
      pointer,
      projectedPoint: pointer,
      edgeMidpoint: pointer,
      edgeLengthMeters: 0,
      edgeWidthMeters: 0,
      distanceMeters: Number.POSITIVE_INFINITY,
      headingDegrees: 0,
      reason: 'SNAP_NO_VALID_EDGE',
    };
  }
  const edge = network.edges.get(snap.edgeId);
  const from = edge ? network.nodes.get(edge.fromNodeId)?.position : undefined;
  const to = edge ? network.nodes.get(edge.toNodeId)?.position : undefined;
  const edgeMidpoint: CityPoint = from && to
    ? { x: (from.x + to.x) / 2, y: (from.y + to.y) / 2, z: (from.z + to.z) / 2 }
    : snap.projectedPoint;
  return {
    accepted: true,
    edgeId: snap.edgeId,
    pointer,
    projectedPoint: snap.projectedPoint,
    edgeMidpoint,
    edgeLengthMeters: edge?.lengthMeters ?? 0,
    edgeWidthMeters: edge ? roadWidthMetersFor(edge, network.dataset.presentation) : 0,
    distanceMeters: snap.distanceMeters,
    headingDegrees: snap.headingDegrees,
    reason: null,
  };
}

/** Placement of every active barrier: snapped point plus the heading that crosses the road. */
export function barrierPlacements(
  snapshot: ScenarioSnapshot,
  network: RoadNetwork = createRoadNetwork(),
  selectedBarrierId: string | null = null,
): BarrierPlacement[] {
  return [...snapshot.barriers]
    .sort((left, right) => (left.barrierId < right.barrierId ? -1 : 1))
    .map((barrier) => {
      const edge = network.edges.get(barrier.blockedEdgeId);
      return {
        barrierId: barrier.barrierId,
        edgeId: barrier.blockedEdgeId,
        position: barrier.position,
        headingDegrees: edge
          ? headingDegrees(edgeDirection(network, barrier.blockedEdgeId))
          : 0,
        selected: barrier.barrierId === selectedBarrierId,
      };
    });
}

export function describeBarrierPreview(preview: BarrierPreview): string {
  if (!preview.accepted || preview.edgeId === null) {
    return `No road within ${SNAP_EDGE_MAX_RADIUS_M.toFixed(0)} m · release to cancel`;
  }
  return (
    `Closure preview · edge ${preview.edgeId} at ${preview.distanceMeters.toFixed(1)} m · ` +
    'blocks both directions'
  );
}

export function describeBarrierPlacement(
  accepted: boolean,
  edgeId: string | null,
  barrierId: string | null,
): string {
  if (!accepted || edgeId === null) {
    return 'Closure rejected · no road edge under the pointer';
  }
  return `Road ${edgeId} closed${barrierId ? ` by ${barrierId}` : ''}`;
}

export function describeBarrierImpactReason(reason: string): string {
  if (reason === 'CLOSED_ROAD') return 'route crosses the closed road';
  if (reason === 'NO_WORK') return 'no stops left after the closure';
  if (reason === 'DELAYED_ORDER') return 'carries an order the plan delays';
  if (reason === 'DELAYED') return 'delayed by the current plan';
  if (reason === 'UNREACHABLE') return 'isolated by the closure';
  return reason.toLowerCase().replaceAll('_', ' ');
}
